crates/astria-composer/src/grpc_collector/mod.rs

```diff
@@ -0,0 +1,127 @@
+// transactions to the ! Executor. The Executor then sends the transactions to the Astria
+// Shared Sequencer.
+
+use std::{
+    net::SocketAddr,
+    time::Duration,
+};
+
+use astria_core::{
+    generated::composer::v1alpha1::{
+        grpc_collector_service_server::{
+            GrpcCollectorService,
+            GrpcCollectorServiceServer,
+        },
+        SubmitRollupTxsRequest,
+    },
+    sequencer::v1::{
+        asset::default_native_asset_id,
+        transaction::action::SequenceAction,
+        RollupId,
+    },
+};
+use astria_eyre::{
+    eyre,
+    eyre::eyre,
+};
+use tokio::{
+    io,
+    net::TcpListener,
+    sync::mpsc::error::SendTimeoutError,
+};
+use tonic::{
+    Request,
+    Response,
+};
+
+use crate::executor::ExecutorHandle;
+
+/// `GrpcCollector` listens for incoming gRPC requests and sends the Rollup transactions to the
+/// Executor. The Executor then sends the transactions to the Astria Shared Sequencer.
+///
+/// It implements the `ComposerService` rpc service.
+///
+/// The composer will only have one `GrpcCollector` running at a time.
+pub(super) struct GrpcCollector {
+    grpc_collector_listener: TcpListener,
+    executor_handle: ExecutorHandle,
+}
+
+impl GrpcCollector {
+    pub(super) fn new(
+        grpc_collector_listener: TcpListener,
+        executor_handle: ExecutorHandle,
+    ) -> Self {
+        Self {
+            grpc_collector_listener,
+            executor_handle,
+        }
+    }
+
+    /// Returns the socker address the grpc collector is served over
+    /// # Errors
+    /// Returns an error if the listener is not bound
+    pub(super) fn grpc_collector_local_addr(&self) -> io::Result<SocketAddr> {
+        self.grpc_collector_listener.local_addr()
+    }
+
+    pub(super) async fn run_until_stopped(self) -> eyre::Result<()> {
+        let composer_service = GrpcCollectorServiceServer::new(self.executor_handle);
+        let grpc_server = tonic::transport::Server::builder().add_service(composer_service);
+
+        match grpc_server
+            .serve_with_incoming(tokio_stream::wrappers::TcpListenerStream::new(
+                self.grpc_collector_listener,
+            ))
+            .await
+        {
+            Ok(()) => Ok(()),
+            Err(err) => Err(eyre!(err)),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl GrpcCollectorService for ExecutorHandle {
+    async fn submit_rollup_txs(
+        &self,
+        request: Request<SubmitRollupTxsRequest>,
+    ) -> Result<Response<()>, tonic::Status> {
+        let submit_rollup_txs_request = request.into_inner();
+        if submit_rollup_txs_request.rollup_txs.is_empty() {
+            return Err(tonic::Status::invalid_argument(
+                "No sequence actions provided",
+            ));
+        }
+
+        // package the sequence actions into a SequenceAction and send it to the searcher
+        for sequence_action in submit_rollup_txs_request.rollup_txs {
```

Comment:
i still find it somewhat unexpected that the txs passed in via grpc are unbundled and sent separately - can you add a comment on this behaviour on the protobuf? do you plan to change this later?

Reply:
Yes, we will make these channel take in a `Vec<SequenceAction>`, rather than just sequence action in a future PR!